src/include/rados/librados.hpp

```diff
@@ -715,11 +715,17 @@ namespace librados
     /**
      * Rewrite the object with the replica hosted by specified osd
      *
-     * @param osd from which OSD we will copy the data
+     * @param bad_shard a list of <osd_id, shard_id> pairs. it is a black list
+     *                  of if the object shards/replicas, we will avoid using
+     *                  shards/replicas from them. if the object specified by @c
+     *                  oid is located in a replicated pool, the shard_id will
+     *                  be ignored.
      * @param version the version of rewritten object
      * @param what the flags indicating what we will copy
      */
-    int repair_copy(const std::string& oid, uint64_t version, uint32_t what, int32_t osd, uint32_t epoch);
+    int repair_copy(const std::string& oid, uint64_t version, uint32_t what,
+                    const std::vector<std::pair<int32_t, int8_t>>& bad_osds,
```

Comment:
Do we need to expose the <osd, shard> tuple anywhere else?  Probably worth exposing a librados type to wrap it.